opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,249 @@
+                    )
+                    await self.connector.connection.room_redact(
+                        room_id=self.room_id, event_id=event["event_id"],
+                    )
+
+        self.should_migrate = False
+
+    async def connect(self):
+        """Connect to the database."""
+
+        _LOGGER.info("Matrix Database connector initialised.")
+
+    async def put(self, key, value):
+        """Insert or replace a value into the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        _LOGGER.debug(f"Putting {key} into matrix room {self.room_id}.")
+
+        ori_data = await self.connector.connection.room_get_state_event(
+            room_id=self.room_id, event_type=self._event_type, state_key=state_key,
+        )
+        if isinstance(ori_data, RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error putting key into matrix room {self.room_id}: {ori_data.message}({ori_data.status_code})"
+            )
+            return
+        ori_data = ori_data.content
+
+        if self._single_state_key:
+            value = {key: value}
+
+        elif not isinstance(value, dict):
+            raise ValueError(
+                "When the matrix database is configured with single_state_key=False, "
+                "the value passed must be a dict."
+            )
+
+        sub_key = list(value.keys())[0]
+
+        if list(value.keys())[0] in ori_data:
+            if self._single_state_key:
+                existing_val = await self.get(sub_key)
+            else:
+                existing_val = await self.get({key: sub_key})
+            if existing_val == value[sub_key]:
+                _LOGGER.debug("Not updating matrix state, as content hasn't changed.")
+                return
+
+        if self.should_encrypt:
+            room_event = await self.opsdroid.send(
+                GenericMatrixRoomEvent(content=value, event_type=self._event_type)
+            )
+            data = {**ori_data, sub_key: {"encrypted_val": room_event.event_id}}
+        else:
+            data = {**ori_data, **value}
+
+        _LOGGER.debug(f"Putting {key} into matrix room {self.room_id} with {data}")
+
+        await self.opsdroid.send(
+            MatrixStateEvent(
+                self._event_type,
+                content=data,
+                target=self.room_id,
+                connector=self.connector,
+                state_key=state_key,
+            )
+        )
+
+        return True
+
+    async def get(self, key):
+        """Get a value from the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        if not self._single_state_key and not isinstance(key, dict):
```

Comment:
I don't think this is correct.

If `single_state_key` is False then we use `key` as the `state_key`, if `single_state_key` is True then we do `value = event["content"][key]`?